[LV][AArch64] LoopVectorizer allows scalable frem instructions

paschalis-mpeis opened the pull request:
In AArch64, when an 'frem' instruction uses scalable vectors, it will be replaced with a vector library call. LoopVectorize is now aware of that so it no longer returns invalid costs.

When it is not scalable, it returns the default costs, which are delegated to the BaseT TTI Implementation.

paschalis-mpeis committed:
[LV][AArch64] LoopVectorizer allows scalable frem instructions

In AArch64, when an 'frem' instruction uses scalable vectors, it will be
replaced with a vector library call. LoopVectorize is now aware of that
so it no longer returns invalid costs.

When it is not scalable, it returns the default costs, which are
delegated to the BaseT TTI Implementation.

mgabka commented on llvm/lib/Analysis/TargetTransformInfo.cpp:
```diff
@@ -17,7 +17,6 @@
 #include "llvm/IR/IntrinsicInst.h"
 #include "llvm/IR/Module.h"
 #include "llvm/IR/Operator.h"
-#include "llvm/IR/PatternMatch.h"
```
looks like accidental change

paschalis-mpeis replied:
This header is not used directly, so I thought of removing it.
It's unrelated to my patch though.

mgabka commented on llvm/test/Analysis/CostModel/AArch64/arith-fp-sve.ll:
```diff
@@ -137,14 +138,14 @@ define void @fdiv() {
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V8F32 = frem <vscale x 8 x float> undef, undef
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V2F64 = frem <vscale x 2 x double> undef, undef
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V4F64 = frem <vscale x 4 x double> undef, undef
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 10 for instruction: %V4F16 = frem <vscale x 4 x half> undef, undef
```
in these tests veclib is not available so the cost should still be invalid

paschalis-mpeis committed:
Addressing reviewers.

Comparing against the default cost, and no longer restricting to
scalable types.

paschalis-mpeis committed:
[LV][AArch64] LoopVectorizer allows scalable frem instructions

LoopVectorizer is aware when a target can replace a scalable frem
instruction with a vector library call and it returns the relevant cost.
Otherwise, it returns an invalid cost (as previously).

Add test that check costs on AArch64, when there is no vector library
available and when there is, with and without tail-folding.

NOTE: Invoking CostModel directly (not through LV) would still return
invalid costs. To avoid this, it would require passing TargetLibraryInfo
to TargetTransformInfo API, which we preferred not to as the 'frem'
instruction is an isolated case.

paschalis-mpeis commented:
Forced pushed a041258 to additionally rebase to main.

CostModel is now computed in LoopVectorizer.

fhahn commented on llvm/test/Analysis/CostModel/AArch64/arith-fp-sve.ll:
```diff
@@ -1,7 +1,8 @@
-; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
-; RUN: opt < %s -enable-no-nans-fp-math -passes="print<cost-model>" 2>&1 -disable-output -mtriple=aarch64 -mattr=+fullfp16 -mattr=+sve | FileCheck %s
+; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -mattr=+sve -mattr=+fullfp16 -enable-no-nans-fp-math -disable-output -passes="print<cost-model>" %s 2>&1 | FileCheck %s
```
unrelated changed?

fhahn commented on llvm/test/Analysis/CostModel/AArch64/arith-fp.ll:
```diff
@@ -1,7 +1,7 @@
 ; RUN: opt < %s -enable-no-nans-fp-math  -passes="print<cost-model>" 2>&1 -disable-output -mtriple=aarch64 -mattr=+fullfp16 | FileCheck %s
 
-target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128"
+target triple = "aarch64-unknown-linux-gnu"
```
unrelated changes?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7142,6 +7142,19 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
+          SmallVector<Type *, 4> OpTypes;
+          for (auto &Op : I->operands())
+            OpTypes.push_back(Op->getType());
+          return TTI.getCallInstrCost(nullptr, VectorTy, OpTypes, CostKind);
```
IIUC this doesn't account for the cost of the `frem` implementation itself. What about the case where a library provides `frem`, but it is more expensive?

paschalis-mpeis replied:
Now it compares against the default cost. Since this is now in place, the restriction to scalable types (in the [next comment](https://github.com/llvm/llvm-project/pull/76247#discussion_r1441661371) is lifted.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7142,6 +7142,19 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
       Op2Info.Kind = TargetTransformInfo::OK_UniformValue;
 
+    // Some targets replace frem with vector library calls.
+    if (I->getOpcode() == Instruction::FRem && VectorTy->isScalableTy()) {
```
Why limit to scalable types?

fhahn commented:
We are already checking if a library function exists, wouldn't it be more accurate to replace to widen recipe with a WidenCall recipe with the function?

mgabka commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -6951,10 +6951,26 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
+    if (I->getOpcode() == Instruction::FRem) {
+      LibFunc Func;
+      if (TLI->getLibFunc(I->getOpcode(), I->getType(), Func)) {
+        if (TLI->isFunctionVectorizable(TLI->getName(Func))) {
```
I think you could just combine this condition with the one above to avoid excessive indentation.

mgabka commented on llvm/test/Analysis/CostModel/AArch64/arith-fp-sve-frem.ll:
```diff
@@ -0,0 +1,98 @@
+; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py UTC_ARGS: --filter "cost.*vscale.*frem" --version 4
+
+; RUN: opt -mattr=+sve -passes=loop-vectorize -debug-only=loop-vectorize -disable-output -S < %s 2>&1 | FileCheck %s --check-prefix=NO-VECLIB
```
could you also addd RUN lines for Neon, i.e without +sve? That wilol require changing the file name as well.

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -6951,10 +6951,26 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
+            OpTypes.push_back(Op->getType());
+          auto CallCost =
+              TTI.getCallInstrCost(nullptr, VectorTy, OpTypes, CostKind);
+          return std::min(InstrCost, CallCost);
```
@mgabka If CallCost is invalid because there's no available mapping or very high because it's suboptimal and the target in question (not AArch64, at least at present) has an frem instruction, we would want to return InstrCost.

If replace-with-veclib were to override that, I think it would be a bug in that pass.

paschalis-mpeis replied:
The below PR will bring such functionality to `replace-with-veclib`. Thank you all for your input!
- #78688

paschalis-mpeis committed:
Considering VF when getting vector call costs.

Added neon tests as well.

mgabka reviewed (approved):
It LGTM, but probably worth to wait some time to give a chance @huntergr-arm and @fhahn to also have a look.

huntergr-arm reviewed (approved):
LGTM
